Add support for from_edge

SarahAlidoost opened the pull request:
closes #5 
closes #20

SarahAlidoost commented:
@fnattino can you please review this PR? The linter errors have been discussed. The function` merge_lines` can be refactored too, but I am not sure if this will be done in the refactoring PR. Another issue is reported at #20.

fnattino reviewed (changes requested):
Hi @SarahAlidoost really great work! I have left few suggestions, but amazing to see you got it working :).

SarahAlidoost commented on README.md:
````diff
@@ -29,3 +29,16 @@ library(rcoins)
+### Load the package in development mode locally
+
+``` r
+library(usethis)
+library(devtools)
+load_all("/path/to/the/rcoins/package")
````
`usethis` is a required package by `devtools`, but this works as well.

SarahAlidoost commented on R/stroke.R:
```diff
@@ -62,6 +67,12 @@ stroke <- function(edges, angle_threshold = 0, attributes = FALSE,
   # verify that the best links identified fulfill input requirements
   final_links <- cross_check_links(best_links, flow_mode)
 
+  # if we are looking for strokes starting from a specific edge, we use
+  # `best_links`
+  if (!is.null(from_edge)) {
+    final_links <- best_links
+  }
+
```
fixed!

SarahAlidoost commented on R/stroke.R:
```diff
@@ -36,7 +36,12 @@ stroke <- function(edges, angle_threshold = 0, attributes = FALSE,
   if (flow_mode) stop("flow mode not implemented.")
-  if (!is.null(from_edge)) stop("from_edge mode not implemented")
+  if (!is.null(from_edge)) {
+    if (!is.list(from_edge)) stop("from_edge must be a list of edge IDs")
```
fixed.

SarahAlidoost commented on R/stroke.R:
```diff
@@ -268,18 +316,26 @@ merge_lines  <- function(nodes, segments, links, from_edge = NULL) {
+
     while (TRUE) {
-      if (is.na(link) || is_segment_used[link]) break
+      # this is becuase segments can be repeated in the strokes when using
+      # from_edge
+      if (is.null(from_edge)) {
+        if (is.na(link) || is_segment_used[link]) break
+      } else {
+        if (is.na(link)) break
+      }
```
fixed.

SarahAlidoost commented on tests/testthat/test-stroke.R:
```diff
@@ -129,22 +129,23 @@ test_that("edges are not split if flow_mode is true", {
-  expected <- sf::st_sfc(sf::st_linestring(c(p1, p2, p5, p6)))
-  actual <- stroke(sfc, flow_mode = FALSE,
-                   from_edge = sf::st_linestring(c(p5, p6)))
+test_that("strokes can be formed starting from a given a list of edge ids", {
+  new_l1 <- sf::st_linestring(c(p1, p2, p3))
+  sfc <- sf::st_sfc(new_l1, l4, l7)
+  stroke_1 <- sf::st_linestring(c(p1, p2, p3))
+  stroke_2 <- sf::st_linestring(c(p1, p2, p5, p6))
+  expected <- sf::st_sfc(stroke_1, stroke_2)
+  actual <- stroke(sfc, flow_mode = FALSE, from_edge = list(1, 3))
```
nice, I implemented your suggested test. However, the second stroke is also reversed i.e. it is sf::st_linestring(c(p6, p5, p2, p1)) now , is this expected? :thinking:

fnattino replied:
Thank you! 

> nice, I implemented your suggested test. However, the second stroke is also reversed i.e. it is sf::st_linestring(c(p6, p5, p2, p1)) now , is this expected? 🤔

Oh good point, I guess it is! Since edge 2 has now p5 as start and p2 as end, it also grows in the opposite direction. Great the test picks this up. :)

SarahAlidoost commented:
> Hi @SarahAlidoost really great work! I have left few suggestions, but amazing to see you got it working :).

Hi @fnattino thanks for reviewing and your suggestions. I addressed the comments. Please let me know if something needs further improvements. Please also check the additional test and my reply [here](https://github.com/CityRiverSpaces/rcoins/pull/19#discussion_r1846626292). Another point, I am wondering if you have seen issue #20.

fnattino commented:
Hi @SarahAlidoost, see #21 for a fix to #20

fnattino reviewed (approved):
Hi @SarahAlidoost, if tests are passing feel free to merge! (no worries about the linting issues)
